cmd/kubectl-rbac-flatten/README.md

```diff
@@ -0,0 +1,113 @@
+a Kubernetes command-line tool, and so can be used as a [kubectl
+plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/).
+
+## Usage
+
+No filesystem I/O is done except that involved in reading the
+kubeconfig file.
+
+The `kubectl-rbac-flatten` command takes all the standard Kubernetes
+client command line flags (e.g., `--kubeconfig`, `--context`) plus the
+additional ones listed below.
+
+```
+      --api-groups strings               comma-separated list of API groups to include; '*' means all (default [*])
+  -o, --output-format string             output format, either json or table (default "table")
+      --resources strings                comma-separated list of resources to include; '*' means all (default [*])
+```
+
+The syntax shown for the default value of `--api-groups` and
+`--resources` is wrong. The correct syntax is a comma-separated
+collection of strings. For example, `--api-groups
+networking.k8s.io,certificates.k8s.io`.
+
+When filtering on API group or "resource", an RBAC rule that matches
+"*" is included in the result.
+
+## Output
+
+Errors are logged to stderr, and as much processing as possible is
+done.
+
+There are two possible output formats, chosen by a command line flag.
+
+### JSON Output
+
+
+The JSON output does a minimal outer product, producing a tuple per
+(Subject, PolicyRule). The tuples also include some additional fields
+describing their source. The precise data type is as follows, where
+`NamespacedName` is imported from `k8s.io/apimachinery/pkg/types`.
+
+```go
+type Tuple struct {
+	Binding  NamespacedName
+	RoleName string
+	Subject  rbac.Subject
+	Rule     rbac.PolicyRule
+}
+```
+
+The output is one JSON value, an array. Each tuple is written on one
+line, with the comma separators on other lines, making it easy to use
+text-based tools for simple searches. For non-trivial searching, you
+can use a real JSON processor.
+
+Following is an example of JSON output.
+
+```json
+[
+{"Binding":{"Namespace":"","Name":"cluster-admin"},"RoleName":"cluster-admin","Subject":{"kind":"Group","apiGroup":"rbac.authorization.k8s.io","name":"system:masters"},"Rule":{"verbs":["*"],"apiGroups":["*"],"resources":["*"]}}
+,
+{"Binding":{"Namespace":"","Name":"cluster-admin"},"RoleName":"cluster-admin","Subject":{"kind":"Group","apiGroup":"rbac.authorization.k8s.io","name":"system:masters"},"Rule":{"verbs":["*"],"nonResourceURLs":["*"]}}
+,
+...
+]
+```
+
+### Tabular Ouptput
+
+The tabular output does a maximal cross product, so that no tuple in
+the result has a list/slice/array-structured field. The tabular output
+has two tables, one for resource-based rules and one for non-resource
+URLs. The second one is omitted if `--api-goups` is given anything
+other than `*`.
+
+The fields in one row are separated by TABs. The empty string API
+group is output as an empty string (thus between two TABs).
+
+The resource-based table omits the column for the role, for brevity.
```

Comment:
Would this omitted column introduce irregularity and confusion to users? People may ask why one table has the column and the other table doesn't.

And the role names contain useful information to users, no?

Reply:
The reason for the irregularity is documented right here.

The binding specifies the role, so omitting loses no information about the "why". Omitting the binding would leave the user guessing, but omitting the role only requires one easy `get` to resolve. The purpose of this table is to help users who already know something about the "what" to discover the "why".

Reply:
I tried to strike a balance between brevity and convenience.

I used as a criterion: will it fit in one line in the widest window that I can fit on my screen, using the default font. That admits ROLE for the second table but not the first.

Reply:
I would put heavy weight on regularity for this multi objective optimization. But looks like the discussion is more about personal preferences, so I'll also respect your choice.

/lgtm